compiler/z/codegen/S390Instruction.cpp

```diff
@@ -1806,6 +1806,35 @@ TR::S390RILInstruction::adjustCallOffsetWithTrampoline(int32_t offset, uint8_t *
    return offsetHalfWords;
    }
 
+void
+TR::S390RILInstruction::addMetaDataForCodeAddress(uint8_t *cursor)
+   {
+   uintptr_t immediateAsAddress = (uintptr_t) (uint32_t) getSourceImmediate();
+   if (getNext()->getOpCode().getOpCodeValue() == TR::InstOpCode::IIHF)
+      immediateAsAddress |= ((uintptr_t) toS390RILInstruction(getNext())->getSourceImmediate()) << 32;
+
+   TR::Compilation *comp = cg()->comp();
+   bool staticPIC = std::find(comp->getStaticPICSites()->begin(), comp->getStaticPICSites()->end(), this) != comp->getStaticPICSites()->end();
+   bool staticMethodPIC = std::find(comp->getStaticMethodPICSites()->begin(), comp->getStaticMethodPICSites()->end(), this) != comp->getStaticMethodPICSites()->end();
+   bool staticHCRPIC = std::find(comp->getStaticHCRPICSites()->begin(), comp->getStaticHCRPICSites()->end(), this) != comp->getStaticHCRPICSites()->end();
+
+   if (staticPIC && immediateAsAddress && !TR::Compiler->cls.sameClassLoaders(comp, (TR_OpaqueClassBlock *) immediateAsAddress, comp->getCurrentMethod()->classOfMethod()))
```

Comment:
In all three cases here, 32 Bit version of the Runtime helper to added location to be patched on class unload and redefinition. Do we handle it correctly? i.e, It seems to be creating the Unloading site and Redefinition site with size 4, I think `compensate` method in the RAT should take care of it but can you summarize how it works with 64-bit address?

Reply:
I've modified the `compensate` methods in https://github.com/eclipse-openj9/openj9/pull/15705 to have a special case for `LLILF`/`IIHF`. After patching the PIC location, it checks if it is in an `LLILF` and the next instruction is an `IIHF`. If it is, it also patches the `IIHF` immediate to hold the upper 32 bits of the new address (in HCR case) or -1 (in unload case).

Reply:
Okk, I think this is potential bug. I can not think of the example right away but we might have subsequent instruction after LLILF that is unrelated, and this change might update that as well. You should set the size parameter correctly when class and instruction address to list of PIC sites, which creates a runtime assumption. Also update the `compensate` in https://github.com/eclipse-openj9/openj9/pull/15705 to account for that.

Reply:
OK I've reverted this PR to draft while I look into how to prevent this bug

Reply:
Together with the update in https://github.com/eclipse-openj9/openj9/pull/15705, this bug should be averted now. The LLILF of the LLILF/IIHF pair is added as a 64 bit pic site (32 bit constants are still just in LLILF and added as 32 bit pic site), and the `compensate` will only update a subsequent IIHF when updating 64 bit pic sites.